File: cs/src/Ice/TcpTransceiver.cs
// **********************************************************************
//
// Copyright (c) 2003-2011 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

namespace IceInternal
{

    using System;
    using System.ComponentModel;
    using System.Diagnostics;
    using System.Net;
    using System.Net.Sockets;

    sealed class TcpTransceiver : Transceiver
    {
        public int initialize()
        {
            if(_state == StateNeedConnect)
            {
                _state = StateConnectPending;
                return SocketOperation.Connect;
            }
            else if(_state <= StateConnectPending)
            {
                try
                {
#if SILVERLIGHT
                    Network.doFinishConnectAsync(_fd, _socketWriteEventArgs);
#else
                    Network.doFinishConnectAsync(_fd, _writeResult);
#endif
                    _writeResult = null;
                    _state = StateConnected;
                    _desc = Network.fdToString(_fd);
                }
                catch(Ice.LocalException ex)
                {
                    if(_traceLevels.network >= 2)
                    {
                        System.Text.StringBuilder s = new System.Text.StringBuilder();
                        s.Append("failed to establish tcp connection\n");
                        s.Append(Network.fdLocalAddressToString(_fd));
                        Debug.Assert(_addr != null);
                        s.Append("\nremote address = " + _addr.ToString() + "\n");
                        s.Append(ex.ToString());
                        _logger.trace(_traceLevels.networkCat, s.ToString());
                    }
                    throw;
                }

                if(_traceLevels.network >= 1)
                {
                    string s = "tcp connection established\n" + _desc;
                    _logger.trace(_traceLevels.networkCat, s);
                }
            }
            Debug.Assert(_state == StateConnected);
            return SocketOperation.None;
        }

        public void close()
        {
            //
            // If the transceiver is not connected, its description is simply "not connected",
            // which isn't very helpful.
            //
            if(_state == StateConnected && _traceLevels.network >= 1)
            {
                string s = "closing tcp connection\n" + ToString();
                _logger.trace(_traceLevels.networkCat, s);
            }

            Debug.Assert(_fd != null);
            try
            {
                _fd.Close();
            }
            catch(SocketException ex)
            {
                throw new Ice.SocketException(ex);
            }
            finally
            {
                _fd = null;
            }
        }

        public bool write(Buffer buf)
        {
#if COMPACT || SILVERLIGHT
            //
            // The Silverlight & Compact .NET Frameworks does not support the use of synchronous socket
            // operations on a non-blocking socket. Returning false here forces the
            // caller to schedule an asynchronous operation.
            //
            return false;
#else
            int packetSize = buf.b.remaining();
            if(AssemblyUtil.platform_ == AssemblyUtil.Platform.Windows)
            {
                //
                // On Windows, limiting the buffer size is important to prevent
                // poor throughput performance when transferring large amounts of
                // data. See Microsoft KB article KB823764.
                //
                if(_maxSendPacketSize > 0 && packetSize > _maxSendPacketSize / 2)
                {
                    packetSize = _maxSendPacketSize / 2;
                }
            }
            else
            {
                if(_blocking > 0)
                {
                    return false;
                }
            }

            while(buf.b.hasRemaining())
            {
                try
                {
                    Debug.Assert(_fd != null);

                    int ret;
                    try
                    {
                        ret = _fd.Send(buf.b.rawBytes(), buf.b.position(), packetSize, SocketFlags.None);
                    }
                    catch(SocketException e)
                    {
                        if(Network.wouldBlock(e))
                        {
                            return false;
                        }
                        throw;
                    }

                    Debug.Assert(ret > 0);

                    if(_traceLevels.network >= 3)
                    {
                        string s = "sent " + ret + " of " + packetSize + " bytes via tcp\n" + ToString();
                        _logger.trace(_traceLevels.networkCat, s);
                    }

                    if(_stats != null)
                    {
                        _stats.bytesSent(type(), ret);
                    }

                    buf.b.position(buf.b.position() + ret);
                    if(packetSize > buf.b.remaining())
                    {
                        packetSize = buf.b.remaining();
                    }
                }
                catch(SocketException ex)
                {
                    if(Network.connectionLost(ex))
                    {
                        throw new Ice.ConnectionLostException(ex);
                    }

                    throw new Ice.SocketException(ex);
                }
            }

            return true; // No more data to send.
#endif
        }

        public bool read(Buffer buf)
        {
#if COMPACT || SILVERLIGHT
            //
            // The Silverlight & Compact .NET Frameworks does not support the use of synchronous socket
            // operations on a non-blocking socket.
            //
            return false;
#else
            // COMPILERFIX: Workaround for Mac OS X broken poll(), see Mono bug #470120
            if(AssemblyUtil.osx_)
            {
                if(_blocking > 0)
                {
                    return false;
                }
            }

            int remaining = buf.b.remaining();
            int position = buf.b.position();

            while(buf.b.hasRemaining())
            {
                try
                {
                    Debug.Assert(_fd != null);

                    int ret;
                    try
                    {
                        //
                        // Try to receive first. Much of the time, this will work and we
                        // avoid the cost of calling Poll().
                        //
                        ret = _fd.Receive(buf.b.rawBytes(), position, remaining, SocketFlags.None);
                        if(ret == 0)
                        {
                            throw new Ice.ConnectionLostException();
                        }
                    }
                    catch(SocketException e)
                    {
                        if(Network.wouldBlock(e))
                        {
                            return false;
                        }
                        throw;
                    }

                    Debug.Assert(ret > 0);

                    if(_traceLevels.network >= 3)
                    {
                        string s = "received " + ret + " of " + remaining + " bytes via tcp\n" + ToString();
                        _logger.trace(_traceLevels.networkCat, s);
                    }

                    if(_stats != null)
                    {
                        _stats.bytesReceived(type(), ret);
                    }

                    remaining -= ret;
                    buf.b.position(position += ret);
                }
                catch(SocketException ex)
                {
                    //
                    // On Mono, calling shutdownReadWrite() followed by read() causes Socket.Receive() to
                    // raise a socket exception with the "interrupted" error code. We need to check the
                    // socket's Connected status before checking for the interrupted case.
                    //
                    if(!_fd.Connected)
                    {
                        throw new Ice.ConnectionLostException(ex);
                    }

                    if(Network.interrupted(ex))
                    {
                        continue;
                    }

                    if(Network.connectionLost(ex))
                    {
                        throw new Ice.ConnectionLostException(ex);
                    }

                    throw new Ice.SocketException(ex);
                }
            }

            return true;
#endif
        }

        public bool startRead(Buffer buf, 
#if SILVERLIGHT
                              IceInternal.ThreadPool.AsyncCallback callback,
#else
                              AsyncCallback callback, 
#endif
                              object state)
        {
            Debug.Assert(_fd != null && _readResult == null);

#if !COMPACT && !SILVERLIGHT
            // COMPILERFIX: Workaround for Mac OS X broken poll(), see Mono bug #470120
            if(AssemblyUtil.osx_)
            {
                if(++_blocking == 1)
                {
                    Network.setBlock(_fd, true);
                }
            }
#endif

            int packetSize = buf.b.remaining();
            if(_maxReceivePacketSize > 0 && packetSize > _maxReceivePacketSize)
            {
                packetSize = _maxReceivePacketSize;
            }

            try
            {
#if SILVERLIGHT
                if(_readResult == null)
                {
                    _readResult = new IceInternal.ThreadPool.IAsyncResult();
                }
                _socketReadEventArgs = new SocketAsyncEventArgs();
                _socketReadEventArgs.RemoteEndPoint = _addr;
                _socketReadEventArgs.Completed += new EventHandler<SocketAsyncEventArgs>(callback);
                _socketReadEventArgs.UserToken = _readResult;
                _readResult.AsyncState = state;
                _readResult.EventArgs = _socketReadEventArgs;
                _socketReadEventArgs.SetBuffer(buf.b.rawBytes(), buf.b.position(), packetSize);
                _readResult.CompletedSynchronously = !_fd.ReceiveAsync(_socketReadEventArgs);
#else
                _readResult = _fd.BeginReceive(buf.b.rawBytes(), buf.b.position(), packetSize, SocketFlags.None, 
                                               callback, state);
#endif
            }
            catch(SocketException ex)
            {
                if(Network.connectionLost(ex))
                {
                    throw new Ice.ConnectionLostException(ex);
                }

                throw new Ice.SocketException(ex);
            }


            return _readResult.CompletedSynchronously;
        }

        public void finishRead(Buffer buf)
        {
            if(_fd == null) // Transceiver was closed
            {
                _readResult = null;
                return;
            }

            Debug.Assert(_fd != null && _readResult != null);

            try
            {
#if SILVERLIGHT
                SocketAsyncEventArgs eventArgs = (SocketAsyncEventArgs)_readResult.EventArgs;
                int ret = 0;
                if(eventArgs.SocketError == SocketError.Success)
                {
                    ret = eventArgs.BytesTransferred;
                }
#else
                int ret = _fd.EndReceive(_readResult);
#endif
                _readResult = null;
                if(ret == 0)
                {
                    throw new Ice.ConnectionLostException();
                }

#if !COMPACT && !SILVERLIGHT
                // COMPILERFIX: Workaround for Mac OS X broken poll(), see Mono bug #470120
                if(AssemblyUtil.osx_)
                {
                    if(--_blocking == 0)
                    {
                        Network.setBlock(_fd, false);
                    }
                }
#endif
                Debug.Assert(ret > 0);

                if(_traceLevels.network >= 3)
                {
                    int packetSize = buf.b.remaining();
                    if(_maxReceivePacketSize > 0 && packetSize > _maxReceivePacketSize)
                    {
                        packetSize = _maxReceivePacketSize;
                    }
                    string s = "received " + ret + " of " + packetSize + " bytes via tcp\n" + ToString();
                    _logger.trace(_traceLevels.networkCat, s);
                }

                if(_stats != null)
                {
                    _stats.bytesReceived(type(), ret);
                }

                buf.b.position(buf.b.position() + ret);
            }
            catch(SocketException ex)
            {
                if(Network.connectionLost(ex))
                {
                    throw new Ice.ConnectionLostException(ex);
                }

                throw new Ice.SocketException(ex);
            }
            catch(ObjectDisposedException ex)
            {
                throw new Ice.ConnectionLostException(ex);
            }
        }

        public bool startWrite(Buffer buf, 
#if SILVERLIGHT
                               IceInternal.ThreadPool.AsyncCallback callback, 
#else
                               AsyncCallback callback, 
#endif
                               object state, out bool completed)
        {
            Debug.Assert(_fd != null && _writeResult == null);

#if !COMPACT && !SILVERLIGHT
            // COMPILERFIX: Workaround for Mac OS X broken poll(), see Mono bug #470120
            if(AssemblyUtil.osx_)
            {
                if(++_blocking == 1)
                {
                    Network.setBlock(_fd, true);
                }
            }
#endif

            if(_state < StateConnected)
            {
#if SILVERLIGHT
                _socketWriteEventArgs = new SocketAsyncEventArgs();
                _writeResult = Network.doConnectAsync(_fd, _addr, callback, _socketWriteEventArgs, state);
#else
                _writeResult = Network.doConnectAsync(_fd, _addr, callback, state);
#endif
                completed = false;
                return _writeResult.CompletedSynchronously;
            }

            //
            // We limit the packet size for beginWrite to ensure connection timeouts are based
            // on a fixed packet size.
            //
            int packetSize = buf.b.remaining();
            if(_maxSendPacketSize > 0 && packetSize > _maxSendPacketSize)
            {
                packetSize = _maxSendPacketSize;
            }

            try
            {
#if SILVERLIGHT
                _socketWriteEventArgs.SetBuffer(buf.b.rawBytes(), buf.b.position(), packetSize);
                _writeResult = new IceInternal.ThreadPool.IAsyncResult();
                _socketWriteEventArgs.UserToken = _writeResult;
                _writeResult.AsyncState = state;
                _writeResult.EventArgs = _socketWriteEventArgs;
                _writeResult.CompletedSynchronously = !_fd.SendAsync(_socketWriteEventArgs);
#else
                _writeResult = _fd.BeginSend(buf.b.rawBytes(), buf.b.position(), packetSize, SocketFlags.None, 
                                             callback, state);
#endif
            }
            catch(SocketException ex)
            {
                if(Network.connectionLost(ex))
                {
                    throw new Ice.ConnectionLostException(ex);
                }

                throw new Ice.SocketException(ex);
            }
            catch(ObjectDisposedException ex)
            {
                throw new Ice.ConnectionLostException(ex);
            }
            
            completed = packetSize == buf.b.remaining();
	        return _writeResult.CompletedSynchronously;
        }

        public void finishWrite(Buffer buf)
        {
            if(_fd == null) // Transceiver was closed
            {
                if(buf.size() - buf.b.position() < _maxSendPacketSize)
                {
                    buf.b.position(buf.size()); // Assume all the data was sent for at-most-once semantics.
                }
                _writeResult = null;
                return;
            }

            Debug.Assert(_fd != null && _writeResult != null);

            if(_state < StateConnected)
            {
                return;
            }

            try
            {
#if SILVERLIGHT
                SocketAsyncEventArgs eventArgs = (SocketAsyncEventArgs)_writeResult.EventArgs;
                int ret = 0;
                if(eventArgs.SocketError == SocketError.Success)
                {
                    ret = eventArgs.BytesTransferred;
                }
#else
                int ret = _fd.EndSend(_writeResult);
#endif
                _writeResult = null;
                if(ret == 0)
                {
                    throw new Ice.ConnectionLostException();
                }
                Debug.Assert(ret > 0);

#if !COMPACT && !SILVERLIGHT
                // COMPILERFIX: Workaround for Mac OS X broken poll(), see Mono bug #470120
                if(AssemblyUtil.osx_)
                {
                    if(--_blocking == 0)
                    {
                        Network.setBlock(_fd, false);
                    }
                }
#endif

                if(_traceLevels.network >= 3)
                {
                    int packetSize = buf.b.remaining();
                    if(_maxSendPacketSize > 0 && packetSize > _maxSendPacketSize)
                    {
                        packetSize = _maxSendPacketSize;
                    }
                    string s = "sent " + ret + " of " + packetSize + " bytes via tcp\n" + ToString();
                    _logger.trace(_traceLevels.networkCat, s);
                }

                if(_stats != null)
                {
                    _stats.bytesSent(type(), ret);
                }

                buf.b.position(buf.b.position() + ret);
            }
            catch(SocketException ex)
            {
                if(Network.connectionLost(ex))
                {
                    throw new Ice.ConnectionLostException(ex);
                }

                throw new Ice.SocketException(ex);
            }
            catch(ObjectDisposedException ex)
            {
                throw new Ice.ConnectionLostException(ex);
            }
        }

        public string type()
        {
            return "tcp";
        }

        public Ice.ConnectionInfo
        getInfo()
        {
            Debug.Assert(_fd != null);
            Ice.TCPConnectionInfo info = new Ice.TCPConnectionInfo();
            EndPoint localEndpoint = Network.getLocalAddress(_fd);
            info.localAddress = Network.endpointAddressToString(localEndpoint);
            info.localPort = Network.endpointPort(localEndpoint);
            EndPoint remoteEndpoint = Network.getRemoteAddress(_fd);
            info.remoteAddress = Network.endpointAddressToString(remoteEndpoint);
            info.remotePort = Network.endpointPort(remoteEndpoint);
            return info;
        }

        public void checkSendSize(Buffer buf, int messageSizeMax)
        {
            if(buf.size() > messageSizeMax)
            {
                Ex.throwMemoryLimitException(buf.size(), messageSizeMax);
            }
        }

        public override string ToString()
        {
            return _desc;
        }

        //
        // Only for use by TcpConnector, TcpAcceptor
        //
        internal TcpTransceiver(Instance instance, Socket fd, EndPoint addr, bool connected)
        {
            _fd = fd;
#if SILVERLIGHT
            _addr = (DnsEndPoint)addr;
#else
            _addr = (IPEndPoint)addr;
#endif
            _traceLevels = instance.traceLevels();
            _logger = instance.initializationData().logger;
            _stats = instance.initializationData().stats;
            _state = connected ? StateConnected : StateNeedConnect;
            _desc = connected ? Network.fdToString(_fd) : "<not connected>";

            _maxSendPacketSize = Network.getSendBufferSize(fd);
            if(_maxSendPacketSize < 512)
            {
                _maxSendPacketSize = 0;
            }

            _maxReceivePacketSize = Network.getRecvBufferSize(fd);
            if(_maxReceivePacketSize < 512)
            {
                _maxReceivePacketSize = 0;
            }
        }

        private Socket _fd;
#if SILVERLIGHT
        private DnsEndPoint _addr;
#else
        private IPEndPoint _addr;
#endif
        private TraceLevels _traceLevels;
        private Ice.Logger _logger;
        private Ice.Stats _stats;
        private string _desc;
        private int _state;
        private int _maxSendPacketSize;
        private int _maxReceivePacketSize;

#if !SILVERLIGHT
        private int _blocking = 0;
#endif	
#if SILVERLIGHT
        private SocketAsyncEventArgs	_socketWriteEventArgs;
        private SocketAsyncEventArgs	_socketReadEventArgs;
        private IceInternal.ThreadPool.IAsyncResult _writeResult;
        private IceInternal.ThreadPool.IAsyncResult _readResult;
#else
        private IAsyncResult _writeResult;
        private IAsyncResult _readResult;
#endif
        private const int StateNeedConnect = 0;
        private const int StateConnectPending = 1;
        private const int StateConnected = 2;
    }

}

File: js/src/Ice/TcpTransceiver.js
// **********************************************************************
//
// Copyright (c) 2003-2013 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

var net = require("net");

var Debug = require("./Debug");
var Ex = require("./Exception");
var ExUtil = require("./ExUtil");
var Network = require("./Network");
var SocketOperation = require("./SocketOperation");

var Conn = require("./Connection").Ice;
var LocalEx = require("./LocalException").Ice;

var StateNeedConnect = 0;
var StateConnectPending = 1;
var StateProxyConnectRequest = 2;
var StateProxyConnectRequestPending = 3;
var StateConnected = 4;
var StateClosed = 5;
var StateError = 6;

var TcpTransceiver = function(instance)
{
    this._traceLevels = instance.traceLevels();
    this._logger = instance.initializationData().logger;
    this._readBuffers = [];
    this._readPosition = 0;
}

TcpTransceiver.createOutgoing = function(instance, addr)
{
    var transceiver = new TcpTransceiver(instance);

    transceiver._fd = new net.Socket();
    transceiver._proxy = instance.networkProxy();
    transceiver._addr = addr;
    transceiver._connected = false;
    transceiver._desc = "remote address: " + Network.addrToString(addr.host, addr.port) + " <not connected>";
    transceiver._state = StateNeedConnect;

    return transceiver;
}

TcpTransceiver.createIncoming = function(instance, fd)
{
    var transceiver = new TcpTransceiver(instance);

    transceiver._fd = fd;
    transceiver._proxy = null;
    transceiver._addr = null;
    transceiver._connected = true;
    transceiver._desc = fdToString(fd);
    transceiver._state = StateConnected;

    return transceiver;
}

TcpTransceiver.prototype.setCallbacks = function(
    connectedCallback,      // function()
    bytesAvailableCallback, // function()
    bytesWrittenCallback,   // function()
    closedCallback,         // function()
    errorCallback           // function(ex)
)
{
    this._connectedCallback = connectedCallback;
    this._bytesAvailableCallback = bytesAvailableCallback;
    this._bytesWrittenCallback = bytesWrittenCallback;
    this._closedCallback = closedCallback;
    this._errorCallback = errorCallback;

    var self = this;
    this._fd.on("connect", function() { self.socketConnected(); });
    this._fd.on("close", function(err) { self.socketClosed(err); });
    this._fd.on("error", function(err) { self.socketError(err); });

    //
    // Don't register for incoming data right away.
    //
    //this._fd.on("data", function(buf) { self.socketBytesAvailable(buf); });
}

//
// Returns SocketOperation.None when initialization is complete.
//
TcpTransceiver.prototype.initialize = function(readBuffer, writeBuffer)
{
    try
    {
        if(this._state === StateNeedConnect)
        {
            this._state = StateConnectPending;
            this._fd.connect(this._addr.port, this._addr.host);
            return SocketOperation.Connect; // Waiting for connect to complete.
        }
        else if(this._state === StateConnectPending)
        {
            //
            // Socket is connected.
            //

            this._desc = fdToString(this._fd, this._proxy, this._addr);

            if(this._proxy !== null)
            {
                //
                // Prepare the read & write buffers in advance.
                //
                this._proxy.beginWriteConnectRequest(this._addr.host, this._addr.port, writeBuffer);
                this._proxy.beginReadConnectRequestResponse(readBuffer);

                //
                // Write the proxy connection message.
                //
                if(this.write(writeBuffer))
                {
                    //
                    // Write completed immediately.
                    //
                    this._proxy.endWriteConnectRequest(writeBuffer);

                    //
                    // Try to read the response.
                    //
                    var dummy = { value: false };
                    if(this.read(readBuffer, dummy))
                    {
                        //
                        // Read completed immediately - fall through.
                        //
                        this._proxy.endReadConnectRequestResponse(readBuffer);
                    }
                    else
                    {
                        //
                        // Return SocketOperation.Read to indicate we need to complete the read.
                        //
                        this._state = StateProxyConnectRequestPending; // Wait for proxy response
                        return SocketOperation.Read;
                    }
                }
                else
                {
                    //
                    // Return SocketOperationWrite to indicate we need to complete the write.
                    //
                    this._state = StateProxyConnectRequest; // Send proxy connect request
                    return SocketOperation.Write;
                }
            }

            this._state = StateConnected;
        }
        else if(this._state === StateProxyConnectRequest)
        {
            //
            // Write completed.
            //
            this._proxy.endWriteConnectRequest(writeBuffer);
            this._state = StateProxyConnectRequestPending; // Wait for proxy response
            return SocketOperation.Read;
        }
        else if(this._state === StateProxyConnectRequestPending)
        {
            //
            // Read completed.
            //
            this._proxy.endReadConnectRequestResponse(readBuffer);
            this._state = StateConnected;
        }
    }
    catch(ex)
    {
        if(ex instanceof Ex.LocalException)
        {
            if(this._traceLevels.network >= 2)
            {
                var s = [];
                s.push("failed to establish tcp connection\n");
                s.push(fdToString(this._fd, this._proxy, this._addr.host, this._addr.port));
                this._logger.trace(this._traceLevels.networkCat, s.join(""));
            }
        }
        throw ex;
    }

    Debug.assert(this._state === StateConnected);
    if(this._traceLevels.network >= 1)
    {
        var s = "tcp connection established\n" + this._desc;
        this._logger.trace(this._traceLevels.networkCat, s);
    }

    return SocketOperation.None;
}

TcpTransceiver.prototype.register = function()
{
    //
    // Register the socket data listener.
    //
    var self = this;
    this._fd.on("data", function(buf) { self.socketBytesAvailable(buf); });
}

TcpTransceiver.prototype.unregister = function()
{
    //
    // Unregister the socket data listener.
    //
    this._fd.removeAllListeners("data");
}

TcpTransceiver.prototype.close = function()
{
    if(this._connected && this._traceLevels.network >= 1)
    {
        this._logger.trace(this._traceLevels.networkCat, "closing " + this.type() + " connection\n" + this._desc);
    }

    Debug.assert(this._fd !== null);
    try
    {
        this._fd.destroy();
    }
    catch(ex)
    {
        throw translateError(ex);
    }
    finally
    {
        this._fd = null;
    }
}

//
// Returns true if all of the data was flushed to the kernel buffer.
//
TcpTransceiver.prototype.write = function(byteBuffer)
{
    var remaining = byteBuffer.remaining;
    Debug.assert(remaining > 0);

    //
    // Create a slice of the source buffer representing the remaining data to be written.
    //
    var slice = byteBuffer.b.slice(byteBuffer.position, byteBuffer.position + remaining);

    //
    // The socket will accept all of the data.
    //
    byteBuffer.position = byteBuffer.position + remaining;

    var self = this;

    var sync = true;
    sync = this._fd.write(slice, null,
            //
            // Callback function invoked when data is eventually written.
            //
            function()
            {
                self.socketBytesWritten(remaining, sync);
            });

    return sync;
}

TcpTransceiver.prototype.read = function(byteBuffer, moreData)
{
    moreData.value = false;

    if(this._readBuffers.length === 0)
    {
        return false; // No data available.
    }

    var avail = this._readBuffers[0].length - this._readPosition;
    Debug.assert(avail > 0);
    var remaining = byteBuffer.remaining;

    while(byteBuffer.remaining > 0)
    {
        if(avail > byteBuffer.remaining)
        {
            avail = byteBuffer.remaining;
        }

        this._readBuffers[0].copy(byteBuffer.b, byteBuffer.position, this._readPosition, this._readPosition + avail);

        byteBuffer.position += avail;
        this._readPosition += avail;
        if(this._readPosition === this._readBuffers[0].length)
        {
            //
            // We've exhausted the current read buffer.
            //
            this._readPosition = 0;
            this._readBuffers.shift();
            if(this._readBuffers.length === 0)
            {
                break; // No more data - we're done.
            }
            else
            {
                avail = this._readBuffers[0].length;
            }
        }
    }

    var n = remaining - byteBuffer.remaining;
    if(n > 0 && this._traceLevels.network >= 3)
    {
        var msg = "received " + n + " of " + remaining + " bytes via " + this.type() + "\n" + this._desc;
        this._logger.trace(this._traceLevels.networkCat, msg);
    }

    moreData.value = this._readBuffers.length > 0;

    return byteBuffer.remaining === 0;
}

TcpTransceiver.prototype.type = function()
{
    return "tcp";
}

TcpTransceiver.prototype.getInfo = function()
{
    Debug.assert(this._fd !== null);
    var info = this.createInfo();
    info.localAddress = this._fd.localAddress;
    info.localPort = this._fd.localPort;
    info.remoteAddress = this._fd.remoteAddress;
    info.remotePort = this._fd.remotePort;
    return info;
}

TcpTransceiver.prototype.createInfo = function()
{
    return new Conn.TCPConnectionInfo();
}

TcpTransceiver.prototype.checkSendSize = function(stream, messageSizeMax)
{
    if(stream.size > messageSizeMax)
    {
        ExUtil.throwMemoryLimitException(stream.size, messageSizeMax);
    }
}

TcpTransceiver.prototype.toString = function()
{
    return this._desc;
}

TcpTransceiver.prototype.socketConnected = function()
{
    this._connected = true;
    this._desc = fdToString(this._fd, this._proxy, this._addr);

    if(this._traceLevels.network >= 1)
    {
        this._logger.trace(this._traceLevels.networkCat, this.type() + " connection established\n" + this._desc);
    }

    Debug.assert(this._connectedCallback !== null);
    this._connectedCallback();
}

TcpTransceiver.prototype.socketBytesAvailable = function(buf)
{
    Debug.assert(this._bytesAvailableCallback !== null);
    //
    // TODO: Should we set a limit on how much data we can read?
    // We can call _fd.pause() to temporarily stop reading.
    //
    if(buf.length > 0)
    {
        this._readBuffers.push(buf);
        this._bytesAvailableCallback();
    }
}

TcpTransceiver.prototype.socketBytesWritten = function(n, sync)
{
    if(this._traceLevels.network >= 3)
    {
        var msg = "sent " + n + " bytes via " + this.type() + "\n" + this._desc;
        this._logger.trace(this._traceLevels.networkCat, msg);
    }

    //
    // Don't invoke the callback if the data was written synchronously.
    //
    if(!sync)
    {
        Debug.assert(this._bytesWrittenCallback !== null);
        this._bytesWrittenCallback();
    }
}

TcpTransceiver.prototype.socketClosed = function(err)
{
    //
    // Don't call the closed callback if an error occurred; the error callback
    // will be called.
    //
    if(!err)
    {
        Debug.assert(this._closedCallback !== null);
        this._closedCallback();
    }
}

TcpTransceiver.prototype.socketError = function(err)
{
    Debug.assert(this._errorCallback !== null);
    this._errorCallback(translateError(err));
}

function fdToString(fd, proxy, targetAddr)
{
    if(fd === null)
    {
        return "<closed>";
    }

    return Network.addressesToString(fd.localAddress, fd.localPort, fd.remoteAddress, fd.remotePort, proxy,
                                     targetAddr);
}

function translateError(ex)
{
    // TODO: Search the exception's error message for symbols like ECONNREFUSED ?
    return new LocalEx.SocketException(0, ex);
}

module.exports = TcpTransceiver;
